Trabalho2/grafo.c

```diff
@@ -10,57 +10,389 @@
 
 grafo* grafo_novo()
 {
-   return NULL;
+    grafo* g = (grafo*)malloc(sizeof(grafo));
+    
+    if(g == NULL){
+        return NULL;
+    }
+
+    g->nos = (no_grafo**) malloc(sizeof(no_grafo));
+    
+    if(g->nos == NULL){
+        free(g);
+        return NULL;
+    }
+
+    g->tamanho = 0;
+
+    return g;
 }
 
 
 void grafo_apaga(grafo* g)
 {
-  
+    if(g == NULL){
+        return;
+    }
+
+    int j = 0;
+
+    if(g->tamanho > 0){
+        for (int i=0;i<g->tamanho;i++)
+        {
+            while(g->nos[i]->tamanho > j)
+            {
+                ligacao* aux = g->nos[i]->ligacoes[j];
+                free(aux);
+                aux = NULL;
+                j++;
+            }
+            j=0;
+            ligacao ** aux2 = g->nos[i]->ligacoes;
+            char *aux3 = g->nos[i]->nome_user;
+            no_grafo *aux4 = g->nos[i];
+                free(aux2);
+                aux2 = NULL;
+                free(aux3);
+                aux3 = NULL;
+                free(aux4);
+                aux4 = NULL;
+        }
+        free(g->nos);
+    }
+    free(g);
 }
 
 
 no_grafo * no_insere(grafo *g, char *user)
 {
+
+    if(g == NULL){
+        return NULL;
+    }
+
+    if(user == NULL){
+        return NULL;
+    }
+
+    int i = 0;
+
+    while (g->tamanho > i)
+    {
+        if(strcmp(g->nos[i]->nome_user, user) == 0){
+            return NULL;
+        }
+        i++;
+    }
     
+    if(g->tamanho > 0){
+        g->nos = (no_grafo**) realloc(g->nos, sizeof(no_grafo)*i);

```

Comment:
```suggestion
        g->nos = (no_grafo**) realloc(g->nos, sizeof(no_grafo)*(g->tamanho+1));
```

This is actually a terrible strategy in terms of performance, but we can't really do better here.

Reply:
notice the sizeof change.

Reply:
I think this is a better solucion, `g->nos = (no_grafo**) realloc(g->nos, sizeof(no_grafo*)*(g->tamanho);`, why do we need the +1?;